ank/src/cli.rs

```diff
@@ -12,22 +12,92 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-use std::error::Error;
+use std::{env, error::Error, str::FromStr};
 
-use clap::{command, Parser, Subcommand};
+use clap::{
+    command,
+    error::{ContextKind, ContextValue, ErrorKind},
+    Parser, Subcommand,
+};
 
 use common::DEFAULT_SERVER_ADDRESS;
 use url::Url;
 
-#[derive(Parser)] // requires `derive` feature
+const ANK_SERVER_URL_ENV_KEY: &str = "ANK_SERVER_URL";
+
+fn create_error_context(
+    cmd: &clap::Command,
+    env_key: &str,
+    arg: Option<&clap::Arg>,
+    arg_value: &String,
+) -> clap::Error {
+    let mut err = clap::Error::new(ErrorKind::ValueValidation).with_cmd(cmd); // the order in which the errors are inserted is important
+    if let Some(arg) = arg {
+        if let Ok(env_value) = env::var(env_key) {
+            if env_value == *arg_value {
+                err.insert(
+                    ContextKind::InvalidArg,
+                    ContextValue::String(format!("environment variable '{}'", env_key)),
+                );
+            } else {
+                err.insert(
+                    ContextKind::InvalidArg,
+                    ContextValue::String(arg.to_string()),
+                );
+            }
+        } else {
+            err.insert(
+                ContextKind::InvalidArg,
+                ContextValue::String(arg.to_string()),
+            );
+        }
+    }
+
+    err.insert(
+        ContextKind::InvalidValue,
+        ContextValue::String(arg_value.clone()),
+    );
+    err
+}
+
+#[derive(Clone)]
+/// Custom url parser as a workaround to Clap's bug about
+/// outputting the wrong error message context
+/// when using the environment variable and not the cli argument.
+/// When using a wrong value inside environment variable
+/// Clap still outputs that the cli argument was wrongly set,
+/// but not the environment variable. This is poor use-ability.
+/// An issue for this bug is already opened (https://github.com/clap-rs/clap/issues/5202).
+/// The code will be removed if the bug in Clap is fixed.
+pub struct ServerUrlParser;
+
+impl clap::builder::TypedValueParser for ServerUrlParser {
+    type Value = Url;
+
+    fn parse_ref(
+        &self,
+        cmd: &clap::Command,
+        arg: Option<&clap::Arg>,
+        value: &std::ffi::OsStr,
+    ) -> Result<Self::Value, clap::Error> {
+        let arg_value = value.to_string_lossy().to_string();
+        let err = create_error_context(cmd, ANK_SERVER_URL_ENV_KEY, arg, &arg_value);
+        let url = Url::from_str(&arg_value).map_err(|_| err)?;
```

Comment:
We lose the reason for the error here. 
I would suggest to create a closure with the error environment at 84 and dynamically build the error providing the closure to the map_err function. You probably won't need the arg_value too when you use the error from the parsing (need to check that).

Reply:
Yes, I add this. I will check what code structure fits best here.

Reply:
As discussed, I have the changes reverted. Because clap does not offer to provide the error source. I have checked several things, but setting the source is private in the inner ErrorType.